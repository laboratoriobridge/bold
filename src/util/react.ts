import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
