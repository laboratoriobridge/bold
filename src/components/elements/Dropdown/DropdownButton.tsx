import * as React from 'react'

import { Omit } from '../../../util/types'
import { Button, ButtonProps } from '../Button'

import { Dropdown, DropdownTargetRenderProps } from './Dropdown'
import { DropdownItem, DropdownItemProps, DropdownMenu } from './DropdownMenu'

export type DropdownItemConfig = Omit<DropdownButtonItemProps, 'closeMenu'>

export interface DropdownButtonProps extends ButtonProps {
    items: DropdownItemConfig[]
}

export class DropdownButton extends React.PureComponent<DropdownButtonProps> {

    render() {
        const { items } = this.props
        return (
            <Dropdown renderTarget={this.renderButton} onChange={this.handleChange}>
                {({ closeMenu, highlightedIndex, getMenuProps, getItemProps }) => (
                    <DropdownMenu highlightedIndex={highlightedIndex} {...getMenuProps()}>
                        {items.map((item, idx) => (
                            <DropdownButtonItem
                                key={idx}
                                closeMenu={closeMenu}
                                {...getItemProps({ item })}
                                highlighted={highlightedIndex === idx}
                                {...item}
                            />
                        ))}
                    </DropdownMenu>
                )}
            </Dropdown>
        )
    }

    renderButton = ({ ref, getRootProps, getToggleButtonProps }: DropdownTargetRenderProps) => {
        const { items, ...other } = this.props
        return (
            <Button
                innerRef={ref}
                {...getRootProps()}
                {...getToggleButtonProps()}
                {...other}
            />
        )
    }

    handleChange = (item: DropdownItemConfig) => {
        item.onSelected && item.onSelected()
    }
}

export interface DropdownButtonItemProps extends DropdownItemProps {
    content: React.ReactNode
    autoClose?: boolean
    closeMenu?(): void
}

export class DropdownButtonItem extends React.PureComponent<DropdownButtonItemProps> {

    static defaultProps: Partial<DropdownButtonItemProps> = {
        content: '',
        autoClose: true,
    }

    render() {
        const { content, closeMenu, autoClose, ...other } = this.props
        return <DropdownItem {...other} onSelected={this.handleClick}>{content}</DropdownItem>
    }

    handleClick = () => {
        const { onSelected, autoClose, closeMenu } = this.props
        if (autoClose) {
            closeMenu()
        }
        onSelected && onSelected()
    }

}
