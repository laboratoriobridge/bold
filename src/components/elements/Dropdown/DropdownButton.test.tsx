import * as React from 'react'
import { fireEvent, render, wait } from 'react-testing-library'

import { withTheme } from '../../../test'

import { DropdownButton, DropdownButtonProps } from './DropdownButton'

const createDropdownButton = (props: Partial<DropdownButtonProps> = {}) => withTheme(
    <DropdownButton
        icon='dots'
        size='small'
        skin='ghost'
        items={[
            { content: <span>Item #1</span> },
            { content: <span>Item #2</span>, onSelected: jest.fn(), autoClose: false },
        ]}
        {...props}
    />
)

describe('DropdownButton', () => {

    it('should render correctly', () => {
        render(createDropdownButton())

        expect(document.body).toMatchSnapshot()
    })

    it('should open the dropdown on click', () => {
        const { container } = render(createDropdownButton())

        const button = container.querySelector('button')

        fireEvent.click(button)

        expect(document.body).toMatchSnapshot()
    })

    it('should allow navigation and select by keyboard', async () => {
        const handleSelectItem1 = jest.fn()
        const handleSelectItem2 = jest.fn()
        const { container } = render(createDropdownButton({
            items: [
                { content: <span>Item #1</span>, onSelected: handleSelectItem1 },
                { content: <span>Item #2</span>, onSelected: handleSelectItem2 },
            ],
        }))

        const button = container.querySelector('button')

        fireEvent.click(button)
        fireEvent.keyDown(button, { key: 'ArrowDown' })
        fireEvent.keyDown(button, { key: 'ArrowDown' })
        fireEvent.keyDown(button, { key: 'Enter' })

        await wait()

        expect(handleSelectItem1).toHaveBeenCalledTimes(0)
        expect(handleSelectItem2).toHaveBeenCalledTimes(1)
    })

})
