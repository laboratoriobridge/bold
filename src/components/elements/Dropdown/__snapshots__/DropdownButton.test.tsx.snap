// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownButton should open the dropdown on click 1`] = `
<body>
  .emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5rem;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #2D343B;
  font-size: 0.875rem;
  padding: calc(0.25rem - 1px) calc(0.5rem - 1px);
}

.emotion-1 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-1 > span > :not(:last-child) {
  margin-right: 0.5rem;
}

.emotion-1:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-1:focus {
  outline: none;
  box-shadow: 0 0 0 2px #0A67D1;
}

.emotion-1:not(:disabled):hover {
  background-color: rgba(45,52,59,0.1);
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<div>
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="close menu"
      aria-labelledby="downshift-1-label"
      aria-owns="downshift-1-menu"
      class="emotion-1"
      data-toggle="true"
      role="button"
      type="button"
    >
      <span>
        <svg
          aria-hidden="true"
          class="emotion-0"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
        >
          <defs>
            <path
              d="M12 8a2 2 0 1 1 .001-4.001A2 2 0 0 1 12 8zm0 2a2 2 0 1 1-.001 4.001A2 2 0 0 1 12 10zm0 6a2 2 0 1 1-.001 4.001A2 2 0 0 1 12 16z"
              id="dots_svg__a"
            />
          </defs>
          <use
            xlink:href="#dots_svg__a"
          />
        </svg>
      </span>
    </button>
  </div>
  .emotion-4 {
  font-weight: bold;
  color: #2D343B;
  white-space: nowrap;
  list-style: none;
  margin: 0;
  padding: 0;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.09);
  border: 1px solid #DCDCE0;
  border-radius: 2px;
  display: inline-block;
  width: auto;
  min-width: 150px;
  background: #FFFFFF;
  text-align: left;
}

.emotion-0 {
  display: block;
  padding: 0.5rem 1rem;
  font-size: 0.875rem;
  color: #2D343B;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover {
  background: #F2F2F7;
}

.emotion-1 {
  margin: 0;
}

.emotion-1 div {
  display: block;
}

.emotion-1:first-of-type a {
  border-top-left-radius: 2px;
  border-top-right-radius: 2px;
}

.emotion-1:last-of-type a {
  border-bottom-left-radius: 2px;
  border-bottom-right-radius: 2px;
}

.emotion-5 {
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  opacity: 0;
  opacity: 1;
}

<div
    class="emotion-5"
    style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; z-index: 1300;"
  >
    <ul
      class="emotion-4"
    >
      <li
        class="emotion-1"
      >
        <a
          class="emotion-0"
        >
          <span>
            Item #1
          </span>
        </a>
      </li>
      <li
        class="emotion-1"
      >
        <a
          class="emotion-0"
        >
          <span>
            Item #2
          </span>
        </a>
      </li>
    </ul>
  </div>
</body>
`;

exports[`DropdownButton should render correctly 1`] = `
.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1.5rem;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #2D343B;
  font-size: 0.875rem;
  padding: calc(0.25rem - 1px) calc(0.5rem - 1px);
}

.emotion-1 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-1 > span > :not(:last-child) {
  margin-right: 0.5rem;
}

.emotion-1:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-1:focus {
  outline: none;
  box-shadow: 0 0 0 2px #0A67D1;
}

.emotion-1:not(:disabled):hover {
  background-color: rgba(45,52,59,0.1);
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<body>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="open menu"
      aria-labelledby="downshift-0-label"
      class="emotion-1"
      data-toggle="true"
      role="button"
      type="button"
    >
      <span>
        <svg
          aria-hidden="true"
          class="emotion-0"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
        >
          <defs>
            <path
              d="M12 8a2 2 0 1 1 .001-4.001A2 2 0 0 1 12 8zm0 2a2 2 0 1 1-.001 4.001A2 2 0 0 1 12 10zm0 6a2 2 0 1 1-.001 4.001A2 2 0 0 1 12 16z"
              id="dots_svg__a"
            />
          </defs>
          <use
            xlink:href="#dots_svg__a"
          />
        </svg>
      </span>
    </button>
  </div>
</body>
`;
